utils/githooks/find_base.sh

```diff
@@ -23,9 +30,23 @@ if command -v gh > /dev/null 2>&1; then
         fi
     fi
 else
-    # With no 'gh' command installed then check against origin/master.
+    # With no 'gh' command installed, use the "closest" branch as the target,
+    # calculated as the sum of the commits this branch is ahead and behind.
+    # check master, then current release branches, then current feature branches.
     # shellcheck disable=SC2034
-    TARGET="$ORIGIN"/master
+    all_bases=("master" "release/2.4" "feature/cat_recovery" "feature/multiprovider")
```

Comment:
I wonder if we could just look at the nearest branch and not try to figure out the base branch at all.   That would be a more general solution.

Reply:
Maybe something similar to this but with extra branches added via a .gitignore file similar to how one can add local githooks ?

Reply:
Part of the problem is `git` doesn't really have the concept of "base branch" or "nearest branch" baked in. A branch effectively just points to a commit, and commits have parent commits. So git has no idea where you branched from.

> extra branches added via a .gitignore file similar to how one can add local githooks

Certainly can be done without too much effort. In fact, we could have something like
```
utils/githooks/bases.txt      # checked in, contains upstream branches
utils/githooks/bases.user.txt # in .gitignore - user specific
```

Reply:
that could be very helpful

Reply:
We do already have the VERSION file, and could call out to python to parse it.

If the version is x.y.x then ignore z, if y is odd then the target is master, else it's 'release/x.y'.  Doesn't work for feature branches however.  I can see the argument that feature branches should be checked against master anyway though.

Reply:
> I wonder if we could just look at the nearest branch and not try to figure out the base branch at all. That would be a more general solution.
> 
> While I appreciate that gh isn't always installed, something I've run into locally, it also may be the case that devs are operating on a mirror and the base branch may not be one of the upstream ones.

And `gh` is only useful for this once a PR has been created that first commit(s) you do locally before you create the PR will not benefit from the use of `gh` here.

Reply:
A post checkout hook could work, assuming we could determine whether the branch was just created. I.e. not every checkout means we need to update the "bases". It's really just when we start a new branch that we want to know where we started from.

Reply:
+1 on `gh`. I haven't even bothered to install because I don't find it useful if I have to create a PR first

Reply:
> We do already have the VERSION file, and could call out to python to parse it.
>
> If the version is x.y.x then ignore z, if y is odd then the target is master, else it's 'release/x.y'. Doesn't work for feature branches however. I can see the argument that feature branches should be checked against master anyway though.

Yeah I think that has the same fundamental problem if it doesn't work for feature branches, or as Jeff mentioned, cases where the base isn't an upstream one.

Reply:
Another use case here is I'm working on a hook to automatically determine test tags based on files modified. So we really need to know which branch will ultimately be the merge target.

Reply:
My $0.02 here.

Test-tags in the most-recent-commit only are a bad idea anyway, and I say that as someone who has a commit hook to forward-copy tags from one commit to another.
If you want advanced features like that it's not unreasonable to go to the effort of installing the right tools, which in this case is `gh`.

Reply:
there's a repo for `gh` which is great as they upgrade it so often.

Reply:
> there's a repo for gh which is great as they upgrade it so often.

Sure.  But that was really the quite minor point of @daltonbohning's explanation.  `gh` has the problem that it doesn't work until a PR is created.  Not to mention that not tying all of the behaviour specific to working with GitHub would be an added bonus.

Reply:
Right - what Brian said :)

Reply:
Another idea is an ENV for the base/target branch, similar to the existing `DAOS_ORIGIN` env. I'm not particularly a fan of ENVs, but it does add the most flexibility.
So we could have some set of these solutions:
1. An ENV to specify base
2. PR's current  solution- hardcoded list of potential bases
3. A user file to specify user-specific potential bases - to be in gitignore, and with appropriate README explanation

I'm personally leaning towards (3) to cover the use case @jolivier23 mentioned.

Reply:
4. Pick up the branch from the version file for releases, 1st commit to feature branches would use the wrong base but feature branches should not be too far behind are targeting master anyway.

Reply:
(4) Is that the behavior we want? For example, `feature/cat_recovery` -> `VERSION` is `2.5.100`, which corresponds to master. So all githooks for branches based on / targeted for `feature/cat_recovery` will be based on master. But then I would think the GHA should be similar for consistency.
I'm not against that. In fact it would help keep feature branches more clean before they get merged to master, and it's simpler than keeping track of many possible bases

Reply:
Oh yeah, we don't want the commit hook to be stricter than GHA.  I have no problem with the code in this PR but I don't think we want to add another file to three containing the version/branch that would need to be updated on branch etc.  Using an env for `DAOS_ORIGIN` seems OK as that doesn't change for a tree so could be set in `.bash_rc` but the same user might be pushing to multiple branches so I don't think that's a good option.

In essence, I think 1. and 3. are best avoided.

Reply:
@jolivier23 @brianjmurrell Are you okay with the current solution in this PR? At least for now it's better than what we have.

Reply:
It does seem better than the status quo.  The post-checkout hook idea seems pretty interesting to me though.